src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs

```diff
@@ -18,24 +21,34 @@
 
 namespace Microsoft.CodeAnalysis.CodeRefactorings
 {
+    using TypeInfo = System.Reflection.TypeInfo;
+
     [Export(typeof(ICodeRefactoringService)), Shared]
     internal class CodeRefactoringService : ICodeRefactoringService
     {
-        private readonly Lazy<ImmutableDictionary<string, Lazy<IEnumerable<CodeRefactoringProvider>>>> _lazyLanguageToProvidersMap;
+        private readonly Lazy<ImmutableDictionary<string, Lazy<ImmutableArray<CodeRefactoringProvider>>>> _lazyLanguageToProvidersMap;
+        private readonly ConditionalWeakTable<IReadOnlyList<AnalyzerReference>, StrongBox<ImmutableArray<CodeRefactoringProvider>>> _projectRefactoringsMap
+             = new ConditionalWeakTable<IReadOnlyList<AnalyzerReference>, StrongBox<ImmutableArray<CodeRefactoringProvider>>>();
```

Comment:
being keyed on IReadOnlyList worries me.  though tracking this down in the actual impl, i can see that should be fine here.

Reply:
Just curious, what's the issue you thought it might have and why it's OK?

---
In reply to: [283430736](https://github.com/dotnet/roslyn/pull/35685#discussion_r283430736) [](ancestors = 283430736)

Reply:
I suppose it all boils down to boxing of `ImmutableArray`. If the CWT users directly pass an ImmutableArray instance, they would each add a different key to the table.

In this case though, they use a shared box assigned in the constructor - [link](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/ProjectInfo.cs,d78dc11fa6de4977,references))